src/twisted/web/server.py

```diff
@@ -878,3 +884,110 @@ def acceptableProtocols(self):
+        """
+
+        def get_http_host(data):
+            """
+            Return the value of the Host header found inside `data` HTTP
+            raw request.
+
+            @param data: The first data chunk received from the remote HTTP client peer.
+            @type data: L{bytes}.
+
+            @return: The value of the Host header. It returns the empty string when the value could not be extracted.
+            @rtype: L{bytes}
+            """
+            parts = data.split(b"\r\n")
+            if len(parts) == 1:
+                return b""
+
+            for part in parts:
+                if not part.lower().startswith(b"host"):
+                    continue
+
+                segments = part.split(b":", 1)
+                if len(segments) != 2:
+                    return b""
+
+                return segments[1].strip()
+
+            return b""
+
+        # We only want to handle the first package.
+        self.dataReceived = self.wrappedProtocol.dataReceived
+
+        host = get_http_host(data)
+
+        if not host:
+            # Just forward the data to the wrapped protocol.
+            return self.dataReceived(data)
+
+        self.transport.write(
+            b"HTTP/1.1 301 Moved Permanently\r\n"
+            b"Location: https://%s\r\n"
+            b"Cache-Control: no-store, no-cache\r\n"
+            b"Connection: close\r\n"
+            b"Content-Length: 8\r\n"
+            b"\r\n"
+            b"redirect" % (host,)
+        )
+        self.transport.loseConnection()
+
+
+class HTTPSSiteWrapper(TLSMemoryBIOFactory):
+    """
+    An HTTPS L{Site} wrapper that handles TLS setup and HTTP to HTTPS redirection.
+    """
+
+    def __init__(self, contextFactory, site):
+        """
+        @param contextFactory: An SSL context factory defining SSL parameters for
+            the SSL server.
+        @type contextFactory: L{twisted.internet.interfaces.IOpenSSLContextFactory}
+
+        @param site: The HTTP that is wrapped.
+        @type site: L{Site}.
+        """
+        if TLSMemoryBIOFactory is object:
+            raise NotImplementedError(
+                "OpenSSL not available. Try `pip install twisted[tls]`."
+            )
+        self._site = site
+        super().__init__(contextFactory, isClient=False, wrappedFactory=self._site)
+
+    def buildProtocol(self, addr):
+        """
+        Create a protocol to handle the HTTP client connection.
+
+        @param addr: The address the connection is coming from.
+        @type addr: A tuple of (L{bytes}, L{int}).
+
+        @return: A protocol to handle the new connection.
+        @rtype: L{twisted.internet.interfaces.IProtocol}
+        """
+        protocol = self.protocol(self, self.wrappedFactory.buildProtocol(addr))
+        return _HTTPSRedirectProtocolWrapper(
+            factory=WrappingFactory(self._site), wrappedProtocol=protocol
+        )
+
+    def __getattr__(self, name):
+        return getattr(self._site, name)
```

Comment:
Is this level of delegation really necessary?

Reply:
I remember that  it was necessary .  But I can remove it, run the tests and see what breaks :)